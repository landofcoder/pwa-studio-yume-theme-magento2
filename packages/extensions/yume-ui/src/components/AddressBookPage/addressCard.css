.root {
    align-content: flex-start;
    border: 2px solid rgb(var(--venia-global-color-gray-400));
    border-radius: 0.375rem;
    display: grid;
    grid-template-columns: 1fr max-content;
    padding: 1.25rem 2rem;
}

.root_updated {
    composes: root;
    animation: flash var(--venia-global-anim-bounce) 640ms 2;
}

.contentContainer {
    display: grid;
    row-gap: 0.5rem;
}

.actionContainer {
    align-content: start;
    display: grid;
    justify-items: start;
    row-gap: 0.5rem;
}

.defaultBadge {
    width: max-content;
    padding: 0.375rem 1.5rem;
    margin-bottom: 0.25rem;
    border: 1px solid rgb(var(--venia-global-color-gray-400));
    border-radius: 0.375rem;
    font-size: 0.75rem;
    font-weight: 600;
}

.name {
    font-weight: 600;
}

.telephone {
    margin-top: 0.5rem;
}

.linkButton {
    composes: root from '../LinkButton/linkButton.css';
    text-decoration: none;
}

.editButton {
    composes: linkButton;
    /** hidden until PWA-634 is completed */
    visibility: hidden;
}

.deleteButton {
    composes: linkButton;
    /** hidden until PWA-635 is completed */
    visibility: hidden;
}

@media (max-width: 640px) {
    .root {
        padding: 1rem;
    }
}

@media (max-width: 384px) {
    .actionLabel {
        display: none;
    }

    .linkButton {
        padding: 0 0.25rem 0.25rem;
    }
}

@keyframes flash {
    0% {
        opacity: 1;
    }
    50% {
        opacity: 0.5;
    }
    100% {
        opacity: 1;
    }
}
