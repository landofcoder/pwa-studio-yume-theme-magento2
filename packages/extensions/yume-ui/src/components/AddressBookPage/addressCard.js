import React from 'react';
import { FormattedMessage } from 'react-intl';
import { shape, string, bool, arrayOf } from 'prop-types';
import { Trash2 as TrashIcon, Edit2 as EditIcon } from 'react-feather';

import { mergeClasses } from '@magento/venia-ui/lib/classify';
import Icon from '@magento/venia-ui/lib/components/Icon';
import defaultClasses from './addressCard.css';
import LinkButton from '../LinkButton';

const AddressCard = props => {
    const { address, classes: propClasses, countryName } = props;

    const {
        city,
        country_code,
        default_shipping,
        firstname,
        lastname,
        postcode,
        region: { region },
        street,
        telephone
    } = address;

    const classes = mergeClasses(defaultClasses, propClasses);

    const streetRows = street.map((row, index) => {
        return (
            <span className={classes.streetRow} key={index}>
                {row}
            </span>
        );
    });

    const defaultBadge = default_shipping ? (
        <span className={classes.defaultBadge}>
            <FormattedMessage
                id={'addressCard.defaultText'}
                defaultMessage={'Default'}
            />
        </span>
    ) : null;

    const nameString = `${firstname} ${lastname}`;
    const additionalAddressString = `${city}, ${region} ${postcode}`;

    const deleteButtonElement = !default_shipping ? (
        <LinkButton
            classes={{ root: classes.deleteButton }}
            onClick={() => console.log('To be completed by PWA-635')}
        >
            <Icon classes={{ icon: null }} size={16} src={TrashIcon} />
            <span className={classes.actionLabel}>
                <FormattedMessage
                    id="addressBookPage.deleteAddress"
                    defaultMessage="Delete"
                />
            </span>
        </LinkButton>
    ) : null;

    return (
        <div className={classes.root}>
            <div className={classes.contentContainer}>
                {defaultBadge}
                <span className={classes.name}>{nameString}</span>
                {streetRows}
                <span className={classes.additionalAddress}>
                    {additionalAddressString}
                </span>
                <span className={classes.country}>
                    {countryName || country_code}
                </span>
                <span className={classes.telephone}>
                    <FormattedMessage
                        id="addressBookPage.telephone"
                        values={{ telephone }}
                    />
                </span>
            </div>
            <div className={classes.actionContainer}>
                <LinkButton
                    classes={{ root: classes.editButton }}
                    onClick={() => console.log('To be completed by PWA-634')}
                >
                    <Icon classes={{ icon: null }} size={16} src={EditIcon} />
                    <span className={classes.actionLabel}>
                        <FormattedMessage
                            id="addressBookPage.editAddress"
                            defaultMessage="Edit"
                        />
                    </span>
                </LinkButton>
                {deleteButtonElement}
            </div>
        </div>
    );
};

export default AddressCard;

AddressCard.propTypes = {
    address: shape({
        city: string,
        country_code: string,
        default_shipping: bool,
        firstname: string,
        lastname: string,
        postcode: string,
        region: shape({
            region_code: string,
            region: string
        }),
        street: arrayOf(string),
        telephone: string
    }).isRequired,
    classes: shape({
        actionContainer: string,
        actionLabel: string,
        additionalAddress: string,
        contentContainer: string,
        country: string,
        defaultBadge: string,
        defaultCard: string,
        deleteButton: string,
        editButton: string,
        flash: string,
        linkButton: string,
        name: string,
        root: string,
        root_updated: string,
        streetRow: string,
        telephone: string
    }),
    countryName: string
};
