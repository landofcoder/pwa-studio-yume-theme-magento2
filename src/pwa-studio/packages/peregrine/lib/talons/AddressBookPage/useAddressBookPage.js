import { useCallback, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useQuery } from '@apollo/client';

import { useAppContext } from '@magento/peregrine/lib/context/app';
import { useUserContext } from '@magento/peregrine/lib/context/user';

/**
 *  A talon to support the functionality of the Address Book page.
 *
 *  @param {Object} props
 *  @param {Object} props.queries - GraphQL queries to be run by the talon.
 *
 *
 *  @returns {Object}   talonProps
 *  @returns {Object}   talonProps.data - The user's address book data.
 *  @returns {Boolean}  talonProps.isLoading - Indicates whether the user's
 *      address book data is loading.
 */
export const useAddressBookPage = props => {
    const {
        queries: { getCustomerAddressesQuery }
    } = props;

    const [
        ,
        {
            actions: { setPageLoading }
        }
    ] = useAppContext();
    const history = useHistory();
    const [{ isSignedIn }] = useUserContext();
    const { data: customerAddressesData, loading } = useQuery(
        getCustomerAddressesQuery,
        {
            skip: !isSignedIn
        }
    );

    // If the user is no longer signed in, redirect to the home page.
    useEffect(() => {
        if (!isSignedIn) {
            history.push('/');
        }
    }, [history, isSignedIn]);

    // Update the page indicator if the GraphQL query is in flight.
    useEffect(() => {
        setPageLoading(loading);
    }, [loading, setPageLoading]);

    const handleAddAddress = useCallback(() => {
        alert('TODO!');
    }, []);

    const customerAddresses =
        (customerAddressesData &&
            customerAddressesData.customer &&
            customerAddressesData.customer.addresses) ||
        [];

    return {
        customerAddresses,
        handleAddAddress
    };
};
