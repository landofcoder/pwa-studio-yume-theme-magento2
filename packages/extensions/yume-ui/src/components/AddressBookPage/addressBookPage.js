import React, { useMemo } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { PlusSquare } from 'react-feather';
import { useAddressBookPage } from '@magento/peregrine/lib/talons/AddressBookPage/useAddressBookPage';

import { mergeClasses } from '@magento/venia-ui/lib/classify';
import { Title } from '@magento/venia-ui/lib/components/Head';
import Icon from '@magento/venia-ui/lib/components/Icon';
import LinkButton from '@magento/venia-ui/lib/components/LinkButton';
import { fullPageLoadingIndicator } from '@magento/venia-ui/lib/components/LoadingIndicator';
import defaultClasses from './addressBookPage.css';
import AddressCard from './addressCard';

const AddressBookPage = props => {
    const talonProps = useAddressBookPage();
    const {
        countryDisplayNameMap,
        customerAddresses,
        handleAddAddress,
        isLoading
    } = talonProps;

    const { formatMessage } = useIntl();
    const classes = mergeClasses(defaultClasses, props.classes);

    const PAGE_TITLE = formatMessage({
        id: 'addressBookPage.addressBookText',
        defaultMessage: 'Address Book'
    });
    const addressBookElements = useMemo(() => {
        const addresses = customerAddresses.map(addressEntry => {
            const countryName = countryDisplayNameMap.get(
                addressEntry.country_code
            );

            return (
                <AddressCard
                    key={addressEntry.id}
                    address={addressEntry}
                    countryName={countryName}
                />
            );
        });

        // sort the collection so the default is first
        return addresses.sort(address =>
            address.props.address.default_shipping ? -1 : 1
        );
    }, [countryDisplayNameMap, customerAddresses]);

    if (isLoading) {
        return fullPageLoadingIndicator;
    }

    // STORE_NAME is injected by Webpack at build time.
    const title = `${PAGE_TITLE} - ${STORE_NAME}`;
    return (
        <div className={classes.root}>
            <Title>{title}</Title>
            <h1 className={classes.heading}>{PAGE_TITLE}</h1>
            <div className={classes.content}>
                {addressBookElements}
                <LinkButton
                    className={classes.addButton}
                    key="addAddressButton"
                    onClick={handleAddAddress}
                >
                    <Icon
                        classes={{
                            icon: classes.addIcon
                        }}
                        size={24}
                        src={PlusSquare}
                    />
                    <span className={classes.addText}>
                        <FormattedMessage
                            id={'addressBookPage.addAddressText'}
                            defaultMessage={'Add an Address'}
                        />
                    </span>
                </LinkButton>
            </div>
        </div>
    );
};

export default AddressBookPage;
