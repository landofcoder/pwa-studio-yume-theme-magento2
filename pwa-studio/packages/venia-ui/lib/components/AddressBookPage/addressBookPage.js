import React, { useMemo } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { PlusSquare } from 'react-feather';

import { useAddressBookPage } from '@magento/peregrine/lib/talons/AddressBookPage/useAddressBookPage';
import { mergeClasses } from '@magento/venia-ui/lib/classify';

import { GET_CUSTOMER_ADDRESSES } from '../CheckoutPage/AddressBook/addressBook.gql';
import AddressCard from '../CheckoutPage/AddressBook/addressCard';
import Icon from '../Icon';
import LinkButton from '../LinkButton';
import { Title } from '../Head';
import defaultClasses from './addressBookPage.css';

const AddressBookPage = props => {
    const talonProps = useAddressBookPage({
        queries: {
            getCustomerAddressesQuery: GET_CUSTOMER_ADDRESSES
        }
    });
    const { customerAddresses, handleAddAddress } = talonProps;

    const { formatMessage } = useIntl();
    const classes = mergeClasses(defaultClasses, props.classes);

    const PAGE_TITLE = formatMessage({
        id: 'addressBookPage.addressBookText',
        defaultMessage: 'Address Book'
    });
    const addressBookElements = useMemo(() => {
        return customerAddresses.map(addressEntry => (
            <AddressCard key={addressEntry.id} address={addressEntry} />
        ));
    }, [customerAddresses]);

    // STORE_NAME is injected by Webpack at build time.
    const title = `${PAGE_TITLE} - ${STORE_NAME}`;
    return (
        <div className={classes.root}>
            <Title>{title}</Title>
            <h1 className={classes.heading}>{PAGE_TITLE}</h1>
            <div className={classes.content}>
                <LinkButton
                    className={classes.addButton}
                    key="addAddressButton"
                    onClick={handleAddAddress}
                >
                    <Icon
                        classes={{
                            icon: classes.addIcon
                        }}
                        size={24}
                        src={PlusSquare}
                    />
                    <span className={classes.addText}>
                        <FormattedMessage
                            id={'addressBookPage.addAddressText'}
                            defaultMessage={'Add an Address'}
                        />
                    </span>
                </LinkButton>
                {addressBookElements}
            </div>
        </div>
    );
};

export default AddressBookPage;
