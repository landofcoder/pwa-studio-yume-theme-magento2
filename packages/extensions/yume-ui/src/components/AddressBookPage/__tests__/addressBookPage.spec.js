import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useAddressBookPage } from '@magento/peregrine/lib/talons/AddressBookPage/useAddressBookPage';

import AddressBookPage from '../addressBookPage';

jest.mock('@magento/venia-ui/lib/classify');

jest.mock('../../Head', () => ({ Title: () => 'Title' }));
jest.mock('../../Icon', () => 'Icon');
jest.mock(
    '@magento/peregrine/lib/talons/AddressBookPage/useAddressBookPage',
    () => {
        return {
            useAddressBookPage: jest.fn()
        };
    }
);
jest.mock('../addressCard', () => 'AddressCard');

const props = {};
const talonProps = {
    countryDisplayNameMap: new Map([['US', 'United States']]),
    customerAddresses: [],
    handleAddAddress: jest.fn().mockName('handleAddAddress'),
    isLoading: false
};

it('renders correctly when there are no existing addresses', () => {
    // Arrange.
    useAddressBookPage.mockReturnValueOnce(talonProps);

    // Act.
    const instance = createTestInstance(<AddressBookPage {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

it('renders loading indicator', () => {
    useAddressBookPage.mockReturnValueOnce({ ...talonProps, isLoading: true });

    const instance = createTestInstance(<AddressBookPage {...props} />);

    expect(instance.toJSON()).toMatchSnapshot();
});

it('renders correctly when there are existing addresses', () => {
    // Arrange.
    const myTalonProps = {
        ...talonProps,
        customerAddresses: [
            { id: 'a', country_code: 'US' },
            { id: 'b', country_code: 'US', default_shipping: true },
            { id: 'c', country_code: 'FR' }
        ]
    };
    useAddressBookPage.mockReturnValueOnce(myTalonProps);

    // Act.
    const instance = createTestInstance(<AddressBookPage {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});
