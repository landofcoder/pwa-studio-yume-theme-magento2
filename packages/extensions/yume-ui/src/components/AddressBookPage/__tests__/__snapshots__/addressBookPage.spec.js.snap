// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are existing addresses 1`] = `
<div
  className="root"
>
  Title
  <h1
    className="heading"
  >
    Address Book
  </h1>
  <div
    className="content"
  >
    <AddressCard
      address={
        Object {
          "country_code": "US",
          "default_shipping": true,
          "id": "b",
        }
      }
      countryName="United States"
    />
    <AddressCard
      address={
        Object {
          "country_code": "US",
          "id": "a",
        }
      }
      countryName="United States"
    />
    <AddressCard
      address={
        Object {
          "country_code": "FR",
          "id": "c",
        }
      }
    />
    <button
      className="addButton"
      disabled={false}
      onClick={[MockFunction handleAddAddress]}
      type="button"
    >
      <span
        className="content"
      >
        <Icon
          classes={
            Object {
              "icon": "addIcon",
            }
          }
          size={24}
          src={[Function]}
        />
        <span
          className="addText"
        >
          <mock-FormattedMessage
            defaultMessage="Add an Address"
            id="addressBookPage.addAddressText"
          />
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders correctly when there are no existing addresses 1`] = `
<div
  className="root"
>
  Title
  <h1
    className="heading"
  >
    Address Book
  </h1>
  <div
    className="content"
  >
    <button
      className="addButton"
      disabled={false}
      onClick={[MockFunction handleAddAddress]}
      type="button"
    >
      <span
        className="content"
      >
        <Icon
          classes={
            Object {
              "icon": "addIcon",
            }
          }
          size={24}
          src={[Function]}
        />
        <span
          className="addText"
        >
          <mock-FormattedMessage
            defaultMessage="Add an Address"
            id="addressBookPage.addAddressText"
          />
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`renders loading indicator 1`] = `
<div
  className="global"
>
  <Icon
    classes={
      Object {
        "icon": "indicator",
      }
    }
    size={64}
    src={[Function]}
  />
  <span
    className="message"
  >
    <mock-FormattedMessage
      defaultMessage="Fetching Data..."
      id="loadingIndicator.message"
    />
  </span>
</div>
`;
